### PR TITLE
Relates to #46. Allow calling extrinsic functions expecting not args

### LINKED ISSUE #46: Error calling extrinsics that do not require any arguments
If it's connect to a Substrate kitties node and then try call the `createKitty` Extrinsic, which doesn't require any arguments, and then click "Call" it returns the following error in TxButton because the params value is an array containing an empty string `[""]`
```
Unhandled Rejection (Error): Extrinsic substratekitties.createKitty expects 0 arguments, got 1.

  32 | setStatus("Sending...");
  33 | 
  34 | // Check if this transaction needs sudo
> 35 | let txExecute = sudo ? tx.sudo(...params) : tx(...params);
     | ^  36 | 
  37 | txExecute.signAndSend(fromParam, ({ status }) => {
  38 |   status.isFinalized ?
```

### PR DESCRIPTION
**Pull request by ltfschoen**
Fixes #46

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ Open http://localhost:3000 to view it in the browser.
 The page will reload if you make edits.
 You will also see any lint errors in the console.
 
+## Configure
+
+Configure Custom Types in src/common.json. See https://polkadot.js.org/api/start/types.extend.html
```
Comment (shawntabrizi) [inline README.md]:
```suggestion
Configure custom types in `src/config/common.json`. See [Extending types](https://polkadot.js.org/api/start/types.extend.html).
```

File: src/substrate-lib/components/TxButton.js [hunk 2]
```diff
@@ -31,8 +31,23 @@ export default function TxButton({
...
+
+    if (!hasArgs) {
+      if (!sudo) {
+        txExecute = tx();
+      } else {
+        txExecute = tx.sudo();
+      }
+    } else {
+      if (!sudo) {
+        txExecute = tx(args);
```
Comment (shawntabrizi) [inline src/substrate-lib/components/TxButton.js]:
```suggestion
        txExecute = tx(...params);
```

File: src/substrate-lib/components/TxButton.js [hunk 3]
```diff
@@ -31,8 +31,23 @@ export default function TxButton({
...
+      } else {
+        txExecute = tx.sudo(args);
```
Comment (shawntabrizi) [inline src/substrate-lib/components/TxButton.js]:
```suggestion
        txExecute = tx.sudo(...params);
```

File: src/substrate-lib/components/TxButton.js [hunk 4]
```diff
@@ -31,8 +31,23 @@ export default function TxButton({
     }
     setStatus("Sending...");
 
-    // Check if this transaction needs sudo
-    let txExecute = sudo ? tx.sudo(...params) : tx(...params);
+    // Check if tx has args
+    let hasArgs = params.length && params[0].length ? params : undefined;
```
Comment (shawntabrizi) [inline src/substrate-lib/components/TxButton.js]:
```suggestion
```

File: src/substrate-lib/components/TxButton.js [hunk 5]
```diff
@@ -31,8 +31,23 @@ export default function TxButton({
     }
     setStatus("Sending...");
 
-    // Check if this transaction needs sudo
-    let txExecute = sudo ? tx.sudo(...params) : tx(...params);
+    // Check if tx has args
```
Comment (shawntabrizi) [inline src/substrate-lib/components/TxButton.js]:
```suggestion
    // Check if tx has params
```

File: src/substrate-lib/components/TxButton.js [hunk 6]
```diff
@@ -31,8 +31,23 @@ export default function TxButton({
...
+    let hasArgs = params.length && params[0].length ? params : undefined;
+    let txExecute;
+
+    if (!hasArgs) {
```
Comment (shawntabrizi) [inline src/substrate-lib/components/TxButton.js]:
```suggestion
    if (!params) {
```

**Comment by shawntabrizi** [comment]
Please commit the following changes to make it actually work.

In general, you can reduce the code further using more of the JS `if/else` as it was before:

```
txExecute = !sudo ? tx() : tx.sudo();
```

**Comment by ltfschoen** [comment]
I tried to apply all your proposed changes, but if I call `createKitty` it gives the following error because the value of params is `[""]`:
```
Unhandled Rejection (Error): Extrinsic network.createNetwork expects 0 arguments, got 1.
```

Screenshot 2019-09-30 at 12 36 02 am

**Comment by shawntabrizi** [comment]
It must mean you are passing something to the function. the default value of `params` is null. You can also just pass `params = []`

**Comment by shawntabrizi** [comment]
I think I get the problem. You are using the extrinsics component rather than building your own. In the extrinsics component, the default value for input is `""`. So really the fix here is in the extrinsics component.

**Comment by shawntabrizi** [comment]
@ltfschoen try now. I made the tx button better so hopefully errors like the ones you are pointing out don't crash the whole UI, and I made it so that an extrinsic sent with no input passes "null" rather than `[""]`

**Comment by ltfschoen** [comment]
Yes, your changes work too, but if you call a function like `setPrice` that expects two parameters, when you pass those parameters it triggers an error in TxButton.js on line `txExecute = !sudo ? tx(...params) : tx.sudo(...params);` as follows: 
```
Error: Extrinsic substratekitties.setPrice expects 2 arguments, got 1.
```
This was unexpected, since the value of `params` was an array with two elements and we're using the spread operator to use its values as function parameters.

So I tried just hard-coding two parameters for use with `setPrice` like so `txExecute = !sudo ? tx("0xead434c346821f4c5cd182d810fa534b1d25bdf5900f0b9d8256bdcf7699269f", 50) : tx.sudo(...params);`, and then called it again, and it didn't give an error saying that the incorrect amount of arguments were provided

**Comment by ltfschoen** [comment]
FYI, it works if I convert it into a comma separated list and then back into an array before using the spread operator:
```
let sillySpread = params.join(",").split(",");
txExecute = !sudo ? tx(...sillySpread) : tx.sudo(...sillySpread);
```

**Comment by shawntabrizi** [comment]
Luke, please keep this to the issue at hand. The problem with multiple parameters is different than the issue you are posting here, and I did not claim to solve it.

You should be creating **custom** components rather than trying to use the extrinsics, which is known to not be fully supporting multiple parameters.

**Pull request merged**